File: mini-eps/mini-ep1/leapfrog.py
import math

n = 500
x0 = t0 = 0
xF = tF = 10
u = [0.0]*n
v = [0.0]*n
a = [0.0]*n
x = [0.0]*n

deltaX = (xF - x0)/(n - 1)
deltaT = deltaX/2;
nT = int(((tF - t0)/deltaT) + 1)
tempos = [0.0]*int(nT)

tempos[0] = 0
tempos[nT-1] = 10
for i in range(1,nT): tempos[i] = tempos[i-1] + deltaT;
for i in range(0, n):
    x[i] = deltaX * i;
    u[i] = math.exp(-1*(x[i] - 5)*(x[i] - 5));
    v[i] = 0;
    a[i] = 0;


for t in tempos:
    if(t == t0): 
        deltaTLeap = deltaT/2
    else:
        deltaTLeap = deltaT
    for i in range(1,n-1):
        a[i] = (u[i-1] + u[i+1] - 2*u[i])/((deltaX)*(deltaX))
    
    for i in range(0,n):
        v[i] = v[i] + (a[i] * deltaTLeap)
        u[i] = u[i] + (v[i] * deltaT)

for i in range(int((n/2)-10), int((n/2)+11)):
    print(f'({x[i]},{u[i]})')

File: mini-eps/mini-ep1/leapfrog.cpp
#include <iostream>
#include <vector>
#include <cmath>

void leapFrog(){
    int n = 500;
    double nT, x0 = 0, t0 = 0,
    xF = 10, tF = 10, deltaX, deltaT, deltaTLeap;
    std::vector<double> u(n), v(n), a(n), x(n);
    
    deltaX = (xF - x0)/(n - 1); 
    deltaT = deltaX/2;
    nT = ((tF - t0)/deltaT) + 1;
    std::vector<double> tempos(nT);
    tempos[0] = 0;
    tempos[nT-1] = 10;
    for(int i = 0; i < n; i++) {
        x[i] = deltaX * i;
        u[i] = std::exp(-1*(x[i] - 5)*(x[i] - 5));
        v[i] = 0;
        a[i] = 0;
    }
    for(int i = 1; i < nT; i++) tempos[i] = i*deltaT;
    for(double t : tempos){
        if(t == t0) deltaTLeap = deltaT/2;
        else deltaTLeap = deltaT;
        for(int i = 1; i < n-1; i++){
            a[i] = (u[i-1] + u[i+1] - 2*u[i])/((deltaX)*(deltaX));
        }
        for(int i = 0; i < n; i++) v[i] = v[i] + (a[i] * deltaTLeap);
        for(int i = 0; i < n; i++) u[i] = u[i] + (v[i] * deltaT);
    }
    for(int i = (n/2) - 10; i <= (n/2)+10; i++)
        std::cout << '(' << x[i] << ',' << u[i] << ")\n";

}

int main(){
    leapFrog();
    return 0;
}
